crates/ty_python_semantic/src/types/class.rs

```diff
@@ -883,27 +884,122 @@ impl<'db> ClassLiteral<'db> {
+                    dunder_new_function.into_bound_method_type(db, self_ty);
+
+                if let Some(new_return_ty) = new_return_ty {
+                    if !new_return_ty.to_meta_type(db).is_subtype_of(db, self_ty) {
+                        return dunder_new_bound_method;
+                    }
+                }
+                Some(dunder_new_bound_method)
+            } else {
+                None
+            };
+
+        let dunder_init_function_symbol = self_ty
+            .member_lookup_with_policy(
+                db,
+                "__init__".into(),
+                MemberLookupPolicy::MRO_NO_OBJECT_FALLBACK
+                    | MemberLookupPolicy::META_CLASS_NO_TYPE_FALLBACK,
+            )
+            .symbol;
+
+        // TODO: should be the concrete value of `Self`
+        let correct_return_type = Type::instance(db, ClassType::NonGeneric(self));
 
-        if let Symbol::Type(Type::FunctionLiteral(dunder_new_method), _) = dunder_new_method {
-            return Some(dunder_new_method.into_bound_method_type(db, self.into()));
+        // If the class defines an `__init__` method, then we synthesize a `__init__` method
+        // that has the same parameters as the `__init__` method after it is bound, and with the return type
+        // of the concrete value of `Self`.
+        let synthesized_dunder_init_callable =
+            if let Symbol::Type(Type::FunctionLiteral(dunder_init_function), _) =
+                dunder_init_function_symbol
+            {
+                let synthesized_signature = |signature: Signature<'db>| {
+                    Signature::new(signature.parameters().clone(), Some(correct_return_type))
+                        .bind_self()
+                };
+
+                let synthesized_dunder_init_signature = match dunder_init_function.signature(db) {
+                    FunctionSignature::Single(signature) => {
+                        CallableType::single(db, synthesized_signature(signature.clone()))
+                    }
+                    FunctionSignature::Overloaded(overloads, _) => CallableType::from_overloads(
+                        db,
+                        overloads
+                            .iter()
+                            .map(Clone::clone)
+                            .map(synthesized_signature),
+                    ),
+                };
+                Some(Type::Callable(synthesized_dunder_init_signature))
+            } else {
+                None
+            };
+
+        match (dunder_new_function, synthesized_dunder_init_callable) {
+            (Some(dunder_new_function), Some(synthesized_dunder_init_callable)) => {
+                UnionType::from_elements(
+                    db,
+                    vec![dunder_new_function, synthesized_dunder_init_callable],
+                )
+            }
+            (Some(dunder_constructor_function), None)
+            | (None, Some(dunder_constructor_function)) => dunder_constructor_function,
```

Comment:
And similarly here I feel like we need to be returning the bound method version of `__new__`?